### PR TITLE
feat(cache): implement flow node caching

### PR DESCRIPTION
**Pull request by uael**
> [!IMPORTANT]
> Implement caching for flow nodes using `quick_cache` to reduce database queries and improve performance.
> 
>   - **Behavior**:
>     - Implement caching for flow nodes using `quick_cache` in `cache.rs`.
>     - Modify `fetch_code()` and `fetch_flow()` in `cache.rs` to cache flow node data.
>     - Update `resolve_module()` in `flows.rs` to use `fetch_code()`.
>     - Update `load_flow_modules()` in `flows.rs` to use `fetch_flow()`.
>     - Update `push()` in `jobs.rs` to use `fetch_flow()`.
>     - Update `spawn_dedicated_workers_for_flow()` in `dedicated_worker.rs` to use `fetch_code()`.
>     - Update `handle_code_execution_job()` in `worker.rs` to use `fetch_code()`.
>   - **Dependencies**:
>     - Add `quick_cache` to `Cargo.toml` and `Cargo.lock`.
>   - **Error Handling**:
>     - Add `Utf8` error variant to `Error` enum in `error.rs`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=windmill-labs%2Fwindmill&utm_source=github&utm_medium=referral)<sup> for 28d3e5aa013aa3eae6f31a22018228dd5a12a4cf. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: backend/windmill-common/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+pub mod flow {
+    use super::*;
+    use crate::{error, flows::FlowNodeId};
+
+    /// Cache directory for windmill server/worker(s) flow nodes.
+    pub const CACHE_DIR: &str = const_format::concatcp!(super::CACHE_DIR, "flow");
+
+    lazy_static::lazy_static! {
+        /// Flow node cache.
+        /// FIXME: This should be a static but [`Cache`] does not have a static constructor.
```
Comment (uael) [inline backend/windmill-common/src/cache.rs]: nit: `a *const constructor`

File: backend/windmill-common/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    where
+        K: Hash,
+        F: Future<Output = error::Result<[Box<str>; N]>>,
+    {
+        // Generate the file path from `root`` path and `key`.
+        let path = Path::new(root).join(format!("{:016x}", {
+            use std::hash::Hasher;
+
+            let mut hasher = std::hash::DefaultHasher::new();
+            key.hash(&mut hasher);
```
Comment (uael) [inline backend/windmill-common/src/cache.rs]: Hash not needed, use id directly

File: backend/windmill-common/src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            if lock.is_empty() { None } else { Some(String::from(lock)) },
+            String::from(code),
+        ))
+    }
+
+    pub async fn fetch_flow(e: impl PgExecutor<'_>, node: FlowNodeId)
+        -> error::Result<Json<Box<RawValue>>>
+    {
+        CACHE.get_or_insert_async(&node, async {
+            fs::get_or_insert_with(CACHE_DIR, node, || async {
```
Comment (rubenfiszel) [inline backend/windmill-common/src/cache.rs]: wouldn't it be a lot more space efficient to store already parsed than as text?

File: backend/windmill-common/src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                .map_err(Into::into)
+                .map(|record| [record.flow.unwrap_or_default()])
+            })
+            .await
+            .map(|[flow]| Val { flow, ..Default::default() })
+        })
+        .await
+        .map(
+            // SAFETY:
+            // - `RawValue` and `str` have the exact same memory layout.
```
Comment (rubenfiszel) [inline backend/windmill-common/src/cache.rs]: why use str at all and not RawValue dircetly?

File: backend/windmill-common/src/cache.rs (hunk 2 above)
Comment (rubenfiszel) [reply backend/windmill-common/src/cache.rs]: passing a "need_hash" as parameter?

File: backend/windmill-common/src/cache.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    lazy_static::lazy_static! {
+        /// Flow node cache.
+        /// FIXME: This should be a static but [`Cache`] does not have a static constructor.
+        pub static ref CACHE: Cache<FlowNodeId, Val> = Cache::new(1000);
+    }
+
+    /// Flow node cache value.
+    #[derive(Debug, Clone, Default)]
+    pub struct Val {
```
Comment (rubenfiszel) [inline backend/windmill-common/src/cache.rs]: maybe we will just write a second cache but we will likely reuse same-tiered caching for normal workspace scripts as well so might make sense to generalize with a type arg?

File: backend/windmill-common/src/cache.rs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+    /// Cache directory for windmill server/worker(s) flow nodes.
+    pub const CACHE_DIR: &str = const_format::concatcp!(super::CACHE_DIR, "script");
+
+    lazy_static::lazy_static! {
+        /// Scripts cache.
+        /// FIXME: This should be a static but [`Cache`] does not have a const constructor.
+        static ref CACHE: Cache<ScriptHash, Val> = Cache::new(1000);
+    }
+
+    /// Flow node cache value.
```
Comment (uael) [inline backend/windmill-common/src/cache.rs]:
```suggestion
    /// Script cache value.
```

**Pull request merged**